### ISSUE TITLE
Write supervised learning as a python driver

### ISSUE THREAD

**Issue by gcarleo** [netket/netket#367]
@ShHsLin if you had time to look into this at some point it would be great

**Comment by ShHsLin** [comment]
@gcarleo Hi, what do you mean by python driver? I did not follow the recent change. Is it because we are shifting the c++ backend to tensorflow or pytorch now and, thus, comes the rewriting of the functions?
It would be nice if you could point out for me what I should read into to get the current status.

**Comment by gcarleo** [comment]
Yes @ShHsLin the idea is that in v3.0 we are removing all C++ code, to make everything more easily interoperable with autograd engines (pytorch, jax, tensorflow). 

Basically, now all high-level drivers (such as vmc, unsupervised learning etc) are written in pure python (refer to the branch v3.0 for these changes). Most of these drivers are derived from netket/abstract_variational_driver.py. For example you can have a look at how netket/_vmc.py is implemented. 

Let me know if you need more info...

**Comment by ShHsLin** [comment]
@gcarleo Sorry for the delay, I was a bit distracted. I would do it this weekend. Two questions so far:
1.) Should I directly work on the neket/supervised.py file? 
2.) Is it the case that methods for taking computing the gradient w.r.t parameters are still not unify at this point? I mean, for example rbm.py seems to still have the derivative hard coded, and torch.py would take general machine define by computational graph and utiilize the auto-diff functionality to take derivative. And two seems to use different function names. Maybe I did not read the code carefully, but maybe a quick answer would help me out here.  Thanks.

**Comment by gcarleo** [comment]
No problem @ShHsLin and thanks for taking the time to do this!
Things have changed quite a bit in this release 3.0, so it's completely understandable if you are a bit lost here. Let me answer your questions: 

1) Yes, you should do everything in netket/supervised.py . I think the easiest is that you derive the class ```Supervised``` from [this](https://github.com/netket/netket/blob/v3.0/netket/abstract_variational_driver.py) (it is explained fairly well what to do in there, but let me know if you have questions and also look at how _vmc.py is done in practice) 

2) The idea is that all machines are conformant to this [interface](https://github.com/netket/netket/blob/v3.0/netket/machine/abstract_machine.py)  and then we use different backends to implement the derivatives (der_log and vector_jacobian_prod) and the forward methods (log_val). The rbm file you saw is just the numpy implementation of the rbm. You can find alternative implementations in [Jax](https://github.com/netket/netket/blob/e51df15415f62d93810144321ebb41c1e22c9c68/netket/machine/jax.py#L247) or Torch that use AD. In practice, we use AD to compute der_log and/or the vector jacobian product, but not to differentiate the loss function. So basically you should write the gradient of the loss function just using der_log or (better) the vector_jacobian function... 

It is a slightly different philosophy than doing AD all over the place, but this allows us to use different AD backends + use better statistical estimators for the gradients than those you might get doing straightforward AD of the loss function.   

Let me know if you have other questions... maybe for the moment I would suggest to implement only the non SR branch of the optimization, since we are still ironing out the details of the SR interface for non-numpy backends.
